devito/nodes.py

```diff
@@ -0,0 +1,273 @@
+    :param limits: Limits for the iteration space, either the loop size or a
+                   tuple of the form (start, finish, stepping).
+    :param offsets: Optional map list of offsets to honour in the loop
+    """
+
+    is_Iteration = True
+
+    _traversable = ['nodes']
+
+    def __init__(self, nodes, dimension, limits, index=None, offsets=None):
+        # Ensure we deal with a list of Expression objects internally
+        nodes = as_tuple(nodes)
+        self.nodes = as_tuple([n if isinstance(n, Node) else Expression(n)
+                               for n in nodes])
+        assert all(isinstance(i, Node) for i in self.nodes)
+
+        self.dim = dimension
+        self.index = index or self.dim.name
+
+        # Generate loop limits
+        if isinstance(limits, Iterable):
+            assert(len(limits) == 3)
+            self.limits = list(limits)
+        else:
+            self.limits = list((0, limits, 1))
+
+        # Replace open limits with variables names
+        if self.limits[1] is None:
+            # FIXME: Add dimension size as variable bound.
+            # Needs further generalisation to support loop blocking.
+            self.limits[1] = IterationBound("%s_size" % self.dim.name, self.dim)
+
+        # Record offsets to later adjust loop limits accordingly
+        self.offsets = [0, 0]
+        for off in (offsets or {}):
+            self.offsets[0] = min(self.offsets[0], int(off))
+            self.offsets[1] = max(self.offsets[1], int(off))
+
+    def __repr__(self):
+        str_expr = "\n\t".join([str(s) for s in self.nodes])
+        return "Iteration<%s; %s>::\n\t%s" % (self.index, self.limits,
+                                              str_expr)
+
+    @property
+    def ccode(self):
+        """Generate C code for the represented stencil loop
+
+        :returns: :class:`cgen.For` object representing the loop
+        """
+        loop_body = [s.ccode for s in self.nodes]
+        loop_init = c.InlineInitializer(
+            c.Value("int", self.index), "%d + %d" % (self.limits[0], -self.offsets[0]))
+        loop_cond = '%s %s %s' % (self.index, '<' if self.limits[2] >= 0 else '>',
+                                  "%s - %d" % (self.limits[1], self.offsets[1]))
+        loop_inc = '%s += %s' % (self.index, self.limits[2])
+        return c.For(loop_init, loop_cond, loop_inc, c.Block(loop_body))
+
+    @property
+    def children(self):
+        """Return the traversable children."""
+        return (self.nodes,)
+
+
+# Utilities
+
+class Timer(Block):
+
+    """Wrap a Node with C-level timers."""
+
+    def __init__(self, lname, gname, body):
+        """
+        Initialize a Timer object.
+
+        :param lname: Timer name in the local scope.
+        :param gname: Name of the global struct tracking all timers.
+        :param body: Timed block of code.
+        """
+        self._name = lname
+        # TODO: need omp master pragma to be thread safe
```

Comment:
The new infrastructure doesn't deal with OpenMP yet, so these question will all be addressed in upcoming PRs.